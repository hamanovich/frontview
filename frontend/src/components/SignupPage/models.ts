import { AddFlashMessageType, User } from './../../propTypes';

export interface SignupProps {
  signup: (user: User) => Promise<any>;
  addFlashMessage: AddFlashMessageType;
  isUserExists: (identifier: string) => Promise<any>;
}

export interface SignupFormState {
  errors: {
    username?: string;
    email?: string;
    errorMsg?: string;
    [key: string]: string | undefined;
  };
  isLoading: boolean;
  invalid: boolean;
}

export interface SignupFormError {
  message?: string;
  response?: {
    data: {
      error: string;
    };
  };
}
