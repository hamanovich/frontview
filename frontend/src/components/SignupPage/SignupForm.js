import React, { Component } from 'react';
import { func } from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import FontAwesome from 'react-fontawesome';

import Button from 'react-bootstrap/lib/Button';
import Form from 'react-bootstrap/lib/Form';

import { TextField } from '../formElements';

import validate from '../../validations/signup';

export class SignupForm extends Component {
  static propTypes = {
    signup: func.isRequired,
    reset: func.isRequired,
    addFlashMessage: func.isRequired,
    isUserExists: func.isRequired,
    handleSubmit: func.isRequired,
  };

  state = {
    errors: {
      username: '',
      email: '',
    },
    isLoading: false,
    invalid: false,
  };

  onSubmit = values => {
    const { addFlashMessage, signup, reset } = this.props;

    this.setState({ errors: {}, isLoading: true });

    signup(values)
      .then(() => {
        addFlashMessage({
          type: 'warn',
          text: 'Please, verify your email to confirm',
        });

        reset();
      })
      .catch(err =>
        this.setState({
          errors: err.response
            ? err.response.data
            : {
                username: '',
                email: '',
                errorMsg: err.message,
              },
          isLoading: false,
        }),
      );
  };

  checkUserExists = e => {
    const { isUserExists } = this.props;
    const { name, value } = e.target;
    const { errors } = this.state;

    if (value !== '') {
      isUserExists(value).then(res => {
        let invalid;

        if (res.data.username) {
          errors[name] = `There is user with such ${name}`;
          invalid = true;
        } else {
          errors[name] = '';
          invalid = false;
        }

        this.setState({ errors, invalid });
      });
    } else {
      errors[name] = '';
      this.setState({ errors });
    }
  };

  render() {
    const { errors, isLoading, invalid } = this.state;
    const { handleSubmit } = this.props;

    return (
      <Form onSubmit={handleSubmit(this.onSubmit)} noValidate>
        <Field
          label="Username*:"
          component={TextField}
          type="text"
          name="username"
          placeholder="Type your nickname"
          onBlur={this.checkUserExists}
          errorState={errors.username}
        />
        <p className="lead">{errors.username}</p>
        <Field
          label="Email*:"
          component={TextField}
          type="email"
          name="email"
          placeholder="Type your email"
          onBlur={this.checkUserExists}
          errorState={errors.email}
        />
        <Field
          label="Password*:"
          component={TextField}
          type="password"
          name="password"
          placeholder="Come up with a password"
        />
        <Field
          label="Confirm your Password*:"
          component={TextField}
          type="password"
          name="passwordConfirmation"
          placeholder="Repeat your password"
        />
        <Button
          type="submit"
          bsStyle="primary"
          bsSize="large"
          disabled={isLoading || invalid || !!(errors.username || errors.email)}>
          Register
          <FontAwesome name="users" />
        </Button>
      </Form>
    );
  }
}

export default reduxForm({
  form: 'SignupForm',
  validate,
})(SignupForm);
