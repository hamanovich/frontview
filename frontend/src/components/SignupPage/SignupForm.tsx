import React, { Component, SyntheticEvent, FormEvent } from 'react';
import { Field, reduxForm, InjectedFormProps } from 'redux-form';
import FontAwesome from 'react-fontawesome';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';

import { TextField } from '../formElements';

import validate from '../../validations/signup';

import { SignupProps } from './models';

type SignupFormState = {
  errors: {
    username?: string;
    email?: string;
    errorMsg?: string;
    [key: string]: string | undefined;
  };
  isLoading: boolean;
  invalid: boolean;
};

type SignupFormError = {
  message?: string;
  response?: {
    data: {
      error: string;
    };
  };
};

export class SignupForm extends Component<
  SignupProps & InjectedFormProps<{}, SignupProps>,
  SignupFormState
> {
  state: SignupFormState = {
    errors: {
      username: '',
      email: '',
    },
    isLoading: false,
    invalid: false,
  };

  private onSubmit = (values: any) => {
    const { addFlashMessage, signup } = this.props;

    this.setState({ errors: {}, isLoading: true });

    signup(values)
      .then(() => {
        addFlashMessage({
          type: 'warn',
          text: 'Verify your email to confirm',
        });
      })
      .catch((err: SignupFormError) =>
        this.setState({
          errors:
            err.response && err.response.data.error
              ? { errorMsg: err.response.data.error }
              : {
                  username: '',
                  email: '',
                  errorMsg: `${err.message}. Please check your internet connection`,
                },
          isLoading: false,
        }),
      );
  };

  private checkUserExists = (e: FormEvent<HTMLInputElement>) => {
    const { isUserExists } = this.props;
    const { name, value } = e.target as HTMLInputElement;
    const { errors } = this.state;

    if (value !== '') {
      isUserExists(value).then(res => {
        let invalid;

        if (res.data.username) {
          errors[name] = `There is user with such ${name}`;
          invalid = true;
        } else {
          errors[name] = '';
          invalid = false;
        }

        this.setState({ errors, invalid });
      });
    } else {
      errors[name] = '';
      this.setState({ errors });
    }
  };

  render() {
    const { errors, isLoading, invalid } = this.state;
    const { handleSubmit } = this.props;

    return (
      <Form onSubmit={handleSubmit(this.onSubmit)} noValidate>
        {errors.errorMsg && <Alert variant="danger">{errors.errorMsg}</Alert>}
        <Field
          label="Username*:"
          component={TextField}
          type="text"
          name="username"
          placeholder="Type your nickname"
          onBlur={this.checkUserExists}
          errorState={errors.username}
        />
        <Field
          label="Email*:"
          component={TextField}
          type="email"
          name="email"
          placeholder="Type your email"
          onBlur={this.checkUserExists}
          errorState={errors.email}
        />
        <Field
          label="Password*:"
          component={TextField}
          type="password"
          name="password"
          placeholder="Come up with a password"
        />
        <Field
          label="Confirm your Password*:"
          component={TextField}
          type="password"
          name="passwordConfirmation"
          placeholder="Repeat your password"
        />
        <Button
          type="submit"
          variant="primary"
          size="lg"
          disabled={isLoading || invalid || !!(errors.username || errors.email)}>
          Register <FontAwesome name="users" />
        </Button>
      </Form>
    );
  }
}

export default reduxForm<{}, SignupProps>({
  form: 'SignupForm',
  validate,
})(SignupForm);
