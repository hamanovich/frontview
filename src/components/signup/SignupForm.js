import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form';

import Button from 'react-bootstrap/lib/Button';
import Form from 'react-bootstrap/lib/Form';

import TextField from '../formElements/TextField';

import validate from '../../validations/signup';

class SignupForm extends Component {
  static propTypes = {
    userSignup: PropTypes.func.isRequired,
    addFlashMessage: PropTypes.func.isRequired,
    isUserExists: PropTypes.func.isRequired
  };

  static contextTypes = {
    router: PropTypes.object.isRequired
  };

  state = {
    username: '',
    email: '',
    password: '',
    passwordConfirmation: '',
    errors: {},
    isLoading: false
  };

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  isValid = () => {
    const errors = validate(this.state);

    if (!errors.isValid) {
      this.setState({ errors });
    }

    return errors.isValid;
  };

  checkUserExists = (e) => {
    const { isUserExists } = this.props;
    const { name, value } = e.target;
    const errors = this.state.errors;

    if (value !== '') {
      isUserExists(value).then((res) => {
        let invalid;

        if (res.data.user) {
          errors[name] = `There is user with such ${name}`;
          invalid = true;
        } else {
          errors[name] = '';
          invalid = false;
        }

        this.setState({ errors, invalid });
      });
    } else {
      errors[name] = '';
      this.setState({ errors });
    }
  };

  onSubmit = (e) => {
    e.preventDefault();

    const { addFlashMessage, userSignup } = this.props;

    if (this.isValid()) {
      this.setState({ errors: {}, isLoading: true });

      userSignup(this.state).then(
        () => {
          addFlashMessage({
            type: 'success',
            text: 'You have signed up successfully'
          });
          this.context.router.history.push('/');
        },
        err => this.setState({ errors: err.response.data, isLoading: false })
      );
    }
  }

  render() {
    const { errors, isLoading, invalid } = this.state;

    return (
      <Form onSubmit={this.onSubmit} noValidate>
        <Field
          label="Username*:"
          component={TextField}
          type="text"
          name="username"
          placeholder="Type your nickname"
          onChange={this.onChange}
          handleBlur={this.checkUserExists}
          errorState={errors.username}
        />

        <p className="lead">{errors.username}</p>

        <Field
          label="Email*:"
          component={TextField}
          type="email"
          name="email"
          placeholder="Type your email"
          onChange={this.onChange}
          handleBlur={this.checkUserExists}
          errorState={errors.email}
        />

        <Field
          label="Password*:"
          component={TextField}
          type="password"
          name="password"
          placeholder="Come up with a password"
          onChange={this.onChange}
        />

        <Field
          label="Confirm your Password*:"
          component={TextField}
          type="password"
          name="passwordConfirmation"
          placeholder="Repeat your password"
          onChange={this.onChange}
        />

        <Button type="submit" bsStyle="primary" bsSize="large" disabled={isLoading || invalid}>Register</Button>
      </Form>
    );
  }
}

export default reduxForm({
  form: 'signup',
  validate
})(SignupForm);
