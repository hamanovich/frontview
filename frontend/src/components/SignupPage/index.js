import React from 'react';
import { func } from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';

import Row from 'react-bootstrap/lib/Row';
import Col from 'react-bootstrap/lib/Col';
import PageHeader from 'react-bootstrap/lib/PageHeader';
import FontAwesome from 'react-fontawesome';

import { signup, isUserExists } from '../../actions/signup';
import { addFlashMessage } from '../../actions/flash';

import SignupForm from './SignupForm';

export const SignupPage = ({ signup, addFlashMessage, isUserExists }) => (
  <Row>
    <Helmet>
      <title>Frontview: Register. Welcome</title>
    </Helmet>
    <Col md={6} mdOffset={3}>
      <PageHeader>
        <FontAwesome name="user-plus" /> Register. Welcome
      </PageHeader>

      <SignupForm signup={signup} addFlashMessage={addFlashMessage} isUserExists={isUserExists} />
    </Col>
  </Row>
);

SignupPage.propTypes = {
  signup: func.isRequired,
  addFlashMessage: func.isRequired,
  isUserExists: func.isRequired,
};

export default connect(
  null,
  {
    signup,
    addFlashMessage,
    isUserExists,
  },
)(SignupPage);
