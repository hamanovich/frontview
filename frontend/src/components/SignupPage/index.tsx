import React, { FC } from 'react';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';

import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import FontAwesome from 'react-fontawesome';

import { signup, isUserExists } from '../../actions/signup';
import { addFlashMessage } from '../../actions/flash';

import SignupForm from './SignupForm';
import { SignupProps } from './models';

export const SignupPage: FC<SignupProps> = ({ signup, addFlashMessage, isUserExists }) => (
  <Container>
    <Helmet>
      <title>Frontview: Register. Welcome</title>
    </Helmet>
    <Row>
      <Col xl={{ span: 6, offset: 3 }} md={{ span: 8, offset: 2 }}>
        <h1>
          <FontAwesome name="user-plus" /> Register. Welcome
        </h1>

        <SignupForm signup={signup} addFlashMessage={addFlashMessage} isUserExists={isUserExists} />
      </Col>
    </Row>
  </Container>
);

export default connect(
  null,
  {
    signup,
    addFlashMessage,
    isUserExists,
  },
)(SignupPage);
