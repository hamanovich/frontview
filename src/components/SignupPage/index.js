import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Row from 'react-bootstrap/lib/Row';
import Col from 'react-bootstrap/lib/Col';
import PageHeader from 'react-bootstrap/lib/PageHeader';
import FontAwesome from 'react-fontawesome';

import { signup, isUserExists } from '../../actions/signup';
import { addFlashMessage } from '../../actions/flash';

import SignupForm from './SignupForm';

const SignupPage = ({ signup, addFlashMessage, isUserExists }) => (
  <Row>
    <Col md={6} mdOffset={3}>
      <PageHeader>
        <FontAwesome name="user-plus" /> Register. Welcome
      </PageHeader>
      <SignupForm
        signup={signup}
        addFlashMessage={addFlashMessage}
        isUserExists={isUserExists}
      />
    </Col>
  </Row>
);

SignupPage.propTypes = {
  signup: PropTypes.func.isRequired,
  addFlashMessage: PropTypes.func.isRequired,
  isUserExists: PropTypes.func.isRequired
};

export default connect(null, { signup, addFlashMessage, isUserExists })(SignupPage);