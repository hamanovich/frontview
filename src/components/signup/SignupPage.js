import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Row from 'react-bootstrap/lib/Row';
import Col from 'react-bootstrap/lib/Col';

import { userSignup, isUserExists } from '../../actions/signup';
import { addFlashMessage } from '../../actions/flashMessages';

import SignupForm from './SignupForm';

class SignupPage extends Component {
  static propTypes = {
    userSignup: PropTypes.func.isRequired,
    addFlashMessage: PropTypes.func.isRequired,
    isUserExists: PropTypes.func.isRequired
  };

  render() {
    const { userSignup, addFlashMessage, isUserExists } = this.props;

    return (
      <Row>
        <Col md={6} mdOffset={3}>
          <h1>Register. Welcome</h1>
          <SignupForm
            userSignup={userSignup}
            addFlashMessage={addFlashMessage}
            isUserExists={isUserExists}
          />
        </Col>
      </Row>
    );
  }
}

export default connect(null, { userSignup, addFlashMessage, isUserExists })(SignupPage);