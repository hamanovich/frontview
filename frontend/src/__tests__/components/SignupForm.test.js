import React from 'react';
import { shallow } from 'enzyme';

import { SignupForm } from '../../components/SignupPage/SignupForm';

describe('<SignupForm/>', () => {
  const username = 'username';
  const signupOk = jest.fn(() => Promise.resolve('Success'));
  const signupFailed = jest.fn(() => Promise.reject(new Error('Failed')));
  const userExist = jest.fn(() => Promise.resolve({ data: { username } }));
  const userDoesntExist = jest.fn(() => Promise.resolve({ data: { username: undefined } }));
  const props = {
    signup: signupOk,
    reset: jest.fn(),
    addFlashMessage: jest.fn(),
    isUserExists: userExist,
    handleSubmit: fn => fn,
  };
  const initialState = {
    errors: {
      username: '',
      email: '',
    },
    isLoading: false,
    invalid: false,
  };
  const user = {
    email: 'test@test.test',
    password: 'test',
    passwordConfirmation: 'test',
    username: 'test',
  };

  const component = shallow(<SignupForm {...props} />);

  it('renders <SignupForm /> component', () => {
    expect(component).toMatchSnapshot();
  });

  describe('when User blurs from username field and user is exists', () => {
    beforeEach(() =>
      component
        .find('Field[name="username"]')
        .simulate('blur', { target: { value: username, name: 'username' } }),
    );

    it('checks whether user exists or not', () => {
      expect(props.isUserExists).toHaveBeenCalled();
    });

    it('checks user exists and state has error message', () => {
      expect(component.state().errors.username).toBe(`There is user with such ${username}`);
      expect(component.state().invalid).toBeTruthy();
    });

    afterEach(() => component.setState(initialState));
  });

  describe('when User blurs from username field and has no user found', () => {
    const component = shallow(<SignupForm {...props} isUserExists={userDoesntExist} />);
    beforeEach(() =>
      component
        .find('Field[name="username"]')
        .simulate('blur', { target: { value: username, name: 'username' } }),
    );

    it("checks user doesn't exist and state has no error message", () => {
      expect(component.state().errors.username).toBe('');
      expect(component.state().invalid).toBeFalsy();
    });

    afterEach(() => component.setState(initialState));
  });

  describe('when user blurs from username field with empty value', () => {
    beforeEach(() =>
      component
        .find('Field[name="username"]')
        .simulate('blur', { target: { value: '', name: 'username' } }),
    );

    it("doesn't check user existance", () => {
      expect(component.state().errors.username).toBe('');
      expect(component.state().invalid).toBeFalsy();
    });
  });

  describe('when user submits the Signup Form and Promise resolves with status 200', () => {
    beforeEach(() => component.find('ForwardRef(Bootstrap(Form))').simulate('submit', user));

    it('invokes onSubmit method with Ok', () => {
      expect(props.signup).toHaveBeenCalledWith(user);
    });
  });

  describe('when user submits the Signup Form and Promise throws an error', () => {
    const component = shallow(<SignupForm {...props} signup={signupFailed} />);

    beforeEach(() => component.find('ForwardRef(Bootstrap(Form))').simulate('submit', user));

    it('adds errorMsg to state', () => {
      expect(component.state().errors.errorMsg).toBe('Failed');
    });
  });
});
