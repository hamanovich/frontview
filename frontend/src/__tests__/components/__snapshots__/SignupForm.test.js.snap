// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupForm/> renders <SignupForm /> component 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={false}
  noValidate={true}
  onSubmit={[Function]}
>
  <Field
    component={[Function]}
    errorState=""
    label="Username*:"
    name="username"
    onBlur={[Function]}
    placeholder="Type your nickname"
    type="text"
  />
  <p
    className="lead"
  />
  <Field
    component={[Function]}
    errorState=""
    label="Email*:"
    name="email"
    onBlur={[Function]}
    placeholder="Type your email"
    type="email"
  />
  <Field
    component={[Function]}
    label="Password*:"
    name="password"
    placeholder="Come up with a password"
    type="password"
  />
  <Field
    component={[Function]}
    label="Confirm your Password*:"
    name="passwordConfirmation"
    placeholder="Repeat your password"
    type="password"
  />
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="large"
    bsStyle="primary"
    disabled={false}
    type="submit"
  >
    Register
    <FontAwesome
      name="users"
    />
  </Button>
</Form>
`;
